lxd/instance/drivers/driver_qemu.go

```diff
@@ -1559,115 +1559,242 @@ func (vm *qemu) generateQemuConfigFile(bus string, devConfs []*deviceConfig.RunC
+		"nvramPath":    vm.nvramPath(),
 	})
 	if err != nil {
 		return "", err
 	}
 
-	err = qemuRNG.Execute(sb, map[string]interface{}{
-		"architecture": vm.architectureName,
+	err = qemuControlSocket.Execute(sb, map[string]interface{}{
+		"path": vm.monitorPath(),
 	})
 	if err != nil {
 		return "", err
 	}
 
-	err = vm.addMemoryConfig(sb)
-	if err != nil {
-		return "", err
+	// getBusAddress is used to create any needed root ports and provide
+	// the bus and address that should be used by a device. It supports
+	// automatically setting up multi-function devices and optimize their use.
+	portNum := 0
+	devNum := 5
+
+	type entry struct {
+		bridgeDev int // Device number on the root bridge.
+		bridgeFn  int // Function number on the root bridge.
+
+		dev string // Existing device name.
+		fn  int    // Function number on the existing device.
 	}
+	entries := map[string]*entry{}
 
-	err = vm.addCPUConfig(sb)
+	var rootPort *entry
+	getRootPort := func() *entry {
+		if rootPort == nil {
+			rootPort = &entry{
+				bridgeDev: devNum,
+			}
+			devNum++
+		} else {
+			if rootPort.bridgeFn == 7 {
+				rootPort.bridgeFn = 0
+				rootPort.bridgeDev = devNum
+				devNum++
+			} else {
+				rootPort.bridgeFn++
+			}
+		}
+
+		return rootPort
+	}
+
+	getBusAddr := func(group string) (string, string, bool) {
+		// FIXME: Need to figure out if ccw needs any bus logic.
+		if bus == "ccw" {
+			return "", "", false
+		}
+
+		// Find a device group if specified.
+		var p *entry
+		if group != "" {
+			var ok bool
+			p, ok = entries[group]
+			if ok {
+				// Check if group is full.
+				if p.fn == 7 {
+					p.fn = 0
+					if bus == "pci" {
+						p.bridgeDev = devNum
+						devNum++
+					} else if bus == "pcie" {
+						r := getRootPort()
+						p.bridgeDev = r.bridgeDev
+						p.bridgeFn = r.bridgeFn
+					}
+				} else {
+					p.fn++
+				}
+			} else {
+				// Create a new group.
+				p = &entry{}
+
+				if bus == "pci" {
+					p.bridgeDev = devNum
+					devNum++
+				} else if bus == "pcie" {
+					r := getRootPort()
+					p.bridgeDev = r.bridgeDev
+					p.bridgeFn = r.bridgeFn
+				}
+
+				entries[group] = p
+			}
+		} else {
+			// Create a new temporary group.
+			p = &entry{}
+
+			if bus == "pci" {
+				p.bridgeDev = devNum
+				devNum++
+			} else if bus == "pcie" {
+				r := getRootPort()
+				p.bridgeDev = r.bridgeDev
+				p.bridgeFn = r.bridgeFn
+			}
+		}
+
+		multi := p.fn == 0 && group != ""
+
+		if bus == "pci" {
+			return "pci.0", fmt.Sprintf("%x.%d", p.bridgeDev, p.fn), multi
+		}
+
+		if bus == "pcie" {
+			if p.fn == 0 {
+				qemuPCIe.Execute(sb, map[string]interface{}{
+					"index":         portNum,
+					"addr":          fmt.Sprintf("%x.%d", p.bridgeDev, p.bridgeFn),
+					"multifunction": p.bridgeFn == 0,
+				})
+				p.dev = fmt.Sprintf("qemu_pcie%d", portNum)
+				portNum++
+			}
+
+			return p.dev, fmt.Sprintf("00.%d", p.fn), multi
+		}
+
+		return "", "", false
+	}
```

Comment:
It would be nice to move this into its own type struct, like `qemuCmd`, perhaps call it `qemuDevPort`? Then the `entries` and `getRootPort` could be unexported, and it would just export `getBusAddr`.

Also the `get` prefix on `getBusAddr` is somewhat misleading, as it actually mutates both the `entries` map and in some cases the qemu config being generated (as getters are usually non-mutating). Perhaps `qemuDevPort.AllocateBusAddr()` would be clearer?

Reply:
Renaming to `allocateBusAddr` and `allocateRootPort` for now.

Moving to a struct sounds good though but won't do it just now, want this merged ASAP as master isn't building right now.

Reply:
@stgraber cool thanks, yeah I'm keen for this to be merged too as I need it for the GPU work. Right now I'm focusing on applying the driver override technique to VM NICs.